File: sdk/android/src/main/java/com/giso/tracker/Tracker.java
package com.giso.tracker;

import android.app.Activity;
import android.app.Application;
import android.content.Context;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.view.View;
import android.view.ViewParent;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Android 埋点 SDK 门面（单例）。
 *
 * 事件收敛：10 个标准事件全部由 SDK 控制触发时机——
 * 生命周期（含前台 app_heartbeat）、页面、元素曝光/点击。
 */
public final class Tracker {
    private static volatile Tracker instance;
    private static final long HEARTBEAT_MIN_MS = 15_000L;
    private static final long HEARTBEAT_MAX_MS = 300_000L;

    private final TrackerConfig config;
    private final CommonParams common;
    private final EventQueue queue;
    private final ExposureTracker exposure;
    private final Handler mainHandler = new Handler(Looper.getMainLooper());

    /** View → 元素声明；参数继承沿 View 树向上查找 */
    private final Map<View, ElementMeta> metas = new java.util.WeakHashMap<>();

    private String curPgid = "";
    private Map<String, Object> curPgParams;
    private Map<String, Object> curPgPt;
    private long pageEnterTs;
    private String refPgid = "";
    private String refEid = "";
    private long foregroundTs;
    private long heartbeatIntervalMs;
    private long lastHeartbeatTs;
    private final Runnable heartbeatTick = this::onHeartbeatTick;

    public static synchronized Tracker init(Application app, TrackerConfig config) {
        if (instance == null) {
            instance = new Tracker(app, config);
        }
        return instance;
    }

    public static Tracker get() {
        if (instance == null) throw new IllegalStateException("Tracker.init() first");
        return instance;
    }

    private Tracker(Application app, TrackerConfig config) {
        this.config = config;
        this.heartbeatIntervalMs = clampHeartbeat(config.heartbeatIntervalMs);
        this.common = new CommonParams(app, config);
        this.queue = new EventQueue(app, config);
        this.exposure = new ExposureTracker(
                config.exposureRatio, config.exposureDurationMs, config.exposureMaxPerPage,
                (view, dur, ratio) -> onExposure(view, dur, ratio));
        registerLifecycle(app);
        trackInstallAndLaunch(app);
        fetchRemoteConfig();
    }

    /** 拉取服务端口径配置（/v1/config），失败静默沿用本地默认值 */
    private void fetchRemoteConfig() {
        String url = config.endpoint.replaceFirst("/v1/track/?$", "/v1/config");
        if (url.equals(config.endpoint)) return;
        new Thread(() -> {
            java.net.HttpURLConnection conn = null;
            try {
                conn = (java.net.HttpURLConnection) new java.net.URL(url).openConnection();
                conn.setConnectTimeout(5000);
                conn.setReadTimeout(5000);
                if (conn.getResponseCode() != 200) return;
                byte[] body = readBytes(conn.getInputStream());
                JSONObject c = new JSONObject(new String(body, java.nio.charset.StandardCharsets.UTF_8));
                exposure.updateThresholds(
                        (float) c.optDouble("exposure_ratio", config.exposureRatio),
                        c.optLong("exposure_duration_ms", config.exposureDurationMs),
                        c.optInt("exposure_max_per_page", config.exposureMaxPerPage));
                queue.updateBatching(
                        c.optInt("batch_size", config.batchSize),
                        c.optLong("flush_interval_ms", config.flushIntervalMs));
                if (c.has("heartbeat_interval_ms")) {
                    heartbeatIntervalMs = clampHeartbeat(c.optLong("heartbeat_interval_ms", heartbeatIntervalMs));
                }
            } catch (Exception ignored) {
                // 静默降级
            } finally {
                if (conn != null) conn.disconnect();
            }
        }, "giso-tracker-config").start();
    }

    private static long clampHeartbeat(long ms) {
        if (ms < HEARTBEAT_MIN_MS) return HEARTBEAT_MIN_MS;
        if (ms > HEARTBEAT_MAX_MS) return HEARTBEAT_MAX_MS;
        return ms;
    }

    private void startHeartbeat() {
        stopHeartbeat();
        lastHeartbeatTs = System.currentTimeMillis();
        mainHandler.postDelayed(heartbeatTick, heartbeatIntervalMs);
    }

    private void stopHeartbeat() {
        mainHandler.removeCallbacks(heartbeatTick);
    }

    private void onHeartbeatTick() {
        long now = System.currentTimeMillis();
        long dur = Math.max(0L, now - lastHeartbeatTs);
        lastHeartbeatTs = now;
        JSONObject page = pageContext(null);
        try { page.put("fg_dur", dur); } catch (JSONException ignored) { }
        emit("app_heartbeat", page, null, null);
        mainHandler.postDelayed(heartbeatTick, heartbeatIntervalMs);
    }

    public void setUid(String uid) { common.setUid(uid); }
    public void clearUid() { common.setUid(""); }

    /** 业务设备 ID（历史账号体系兼容）；启动后尽早调用 */
    public void setBizDid(String bizDid) { common.setBizDid(bizDid); }
    public void clearBizDid() { common.setBizDid(""); }

    public void flush() { queue.flush(); }

    // ── 页面 ──────────────────────────────────────────────

    /** Activity.onResume / Fragment.onResume 调用 */
    public void enterPage(String pgid, Map<String, Object> pgParams) {
        enterPage(pgid, pgParams, null);
    }

    /** @param pt 后台下发的透传参数包（如推荐 trace），本页所有事件自动携带 */
    public void enterPage(String pgid, Map<String, Object> pgParams, Map<String, Object> pt) {
        if (!curPgid.isEmpty()) exitPage();
        curPgid = pgid;
        curPgParams = pgParams;
        curPgPt = pt;
        pageEnterTs = System.currentTimeMillis();
        exposure.resetPage();
        emit("page_enter", pageContext(null), null, null, mergePt(curPgPt, null));
    }

    /** Activity.onPause / Fragment.onPause 调用 */
    public void exitPage() {
        if (curPgid.isEmpty()) return;
        long stay = System.currentTimeMillis() - pageEnterTs;
        emit("page_exit", pageContext(stay), null, null, mergePt(curPgPt, null));
        refPgid = curPgid;
        curPgid = "";
        curPgParams = null;
        curPgPt = null;
    }

    // ── 元素 ──────────────────────────────────────────────

    /**
     * 声明元素：自动监测曝光，并接管点击上报（在业务 OnClickListener 之外自动上报）。
     * 容器（如视频卡）bind 一次参数，子元素 bind 后自动继承。
     */
    public void bind(View view, ElementMeta meta) {
        metas.put(view, meta);
        exposure.observe(view);
        hookClick(view);
    }

    public void unbind(View view) {
        metas.remove(view);
    }

    // ── 业务事件 ───────────────────────────────────────────

    public void bizEvent(String code, Map<String, Object> params) {
        bizEvent(code, params, null);
    }

    /** @param pt 后台下发的透传参数包，与页面级透传包合并后上报（本次调用优先） */
    public void bizEvent(String code, Map<String, Object> params, Map<String, Object> pt) {
        JSONObject biz = new JSONObject();
        try {
            biz.put("code", code);
            JSONObject p = TrackEvent.mapToJson(params);
            if (p != null) biz.put("params", p);
        } catch (JSONException ignored) { }
        emit("biz_event", pageContext(null), null, biz, mergePt(curPgPt, TrackEvent.mapToJson(pt)));
    }

    // ── 内部：曝光/点击 ────────────────────────────────────

    private void onExposure(View view, long dur, float ratio) {
        ElCtx ctx = elementContext(view);
        if (ctx == null) return;
        try {
            ctx.el.put("exp_dur", dur);
            ctx.el.put("exp_ratio", ratio);
        } catch (JSONException ignored) { }
        emit("element_exposure", pageContext(null), ctx.el, null, mergePt(curPgPt, ctx.pt));
    }

    private void hookClick(View view) {
        // 触摸抬起即上报「点击手势完成」，不依赖业务 listener，不改变事件分发
        view.setOnTouchListener((v, ev) -> {
            if (ev.getAction() == android.view.MotionEvent.ACTION_UP && v.isPressed()) {
                ElCtx ctx = elementContext(v);
                if (ctx != null) {
                    refEid = metas.get(v) != null ? metas.get(v).eid : "";
                    emit("element_click", pageContext(null), ctx.el, null, mergePt(curPgPt, ctx.pt));
                }
            }
            return false;
        });
    }

    private static final class ElCtx {
        JSONObject el;
        JSONObject pt;
    }

    /** 沿 View 树向上收集祖先 bind 元素：参数继承（params 与 pt 同规则）+ mod 推导 */
    private ElCtx elementContext(View view) {
        ElementMeta meta = metas.get(view);
        if (meta == null) return null;

        Map<String, Object> merged = new LinkedHashMap<>();
        Map<String, Object> mergedPt = new LinkedHashMap<>();
        String mod = null;
        java.util.List<ElementMeta> chain = new java.util.ArrayList<>();
        ViewParent p = view.getParent();
        while (p instanceof View) {
            ElementMeta m = metas.get(p);
            if (m != null) {
                chain.add(m);
                if (mod == null) mod = m.eid;
            }
            p = ((View) p).getParent();
        }
        // 自根向叶合并，叶子（自身）优先级最高
        for (int i = chain.size() - 1; i >= 0; i--) {
            if (chain.get(i).params != null) merged.putAll(chain.get(i).params);
            if (chain.get(i).pt != null) mergedPt.putAll(chain.get(i).pt);
        }
        if (meta.params != null) merged.putAll(meta.params);
        if (meta.pt != null) mergedPt.putAll(meta.pt);

        ElCtx ctx = new ElCtx();
        ctx.el = new JSONObject();
        try {
            ctx.el.put("eid", meta.eid);
            if (mod != null) ctx.el.put("mod", mod);
            if (meta.pos != null) ctx.el.put("pos", meta.pos);
            if (!merged.isEmpty()) ctx.el.put("params", TrackEvent.mapToJson(merged));
        } catch (JSONException ignored) { }
        if (!mergedPt.isEmpty()) ctx.pt = TrackEvent.mapToJson(mergedPt);
        return ctx;
    }

    /** 合并页面级与元素/调用级透传包（后者优先）；皆空返回 null */
    private static JSONObject mergePt(Map<String, Object> pagePt, JSONObject overlay) {
        JSONObject base = TrackEvent.mapToJson(pagePt);
        if (base == null || base.length() == 0) return overlay;
        if (overlay == null || overlay.length() == 0) return base;
        try {
            for (java.util.Iterator<String> it = overlay.keys(); it.hasNext(); ) {
                String k = it.next();
                base.put(k, overlay.get(k));
            }
        } catch (JSONException ignored) { }
        return base;
    }

    // ── 内部：生命周期 ─────────────────────────────────────

    private void trackInstallAndLaunch(Context context) {
        SharedPreferences sp = context.getSharedPreferences("giso_tracker", Context.MODE_PRIVATE);
        if (!sp.getBoolean("activated", false)) {
            sp.edit().putBoolean("activated", true).apply();
            emit("app_install", null, null, null);
        }
        emit("app_launch", null, null, null);
    }

    private void registerLifecycle(Application app) {
        app.registerActivityLifecycleCallbacks(new Application.ActivityLifecycleCallbacks() {
            private int started = 0;
            private boolean appForeground = false;

            @Override public void onActivityStarted(Activity a) {
                if (started++ == 0) {
                    appForeground = true;
                    common.onForeground();
                    foregroundTs = System.currentTimeMillis();
                    queue.onForeground(
                            TrackEvent.of("app_foreground", common.snapshot(), null, null, null),
                            3000L);
                    startHeartbeat();
                }
            }

            @Override public void onActivityStopped(Activity a) {
                if (--started == 0 && appForeground) {
                    appForeground = false;
                    stopHeartbeat();
                    long fgDur = System.currentTimeMillis() - foregroundTs;
                    JSONObject page = pageContext(null);
                    try { page.put("fg_dur", fgDur); } catch (JSONException ignored) { }
                    // 主线程同步 flush + WakeLock；超时加长以应对弱网 / OEM
                    queue.onBackground(
                            TrackEvent.of("app_background", common.snapshot(), page, null, null),
                            5000L);
                }
            }

            @Override public void onActivityCreated(Activity a, Bundle b) { }
            @Override public void onActivityResumed(Activity a) { common.touch(); }
            @Override public void onActivityPaused(Activity a) { }
            @Override public void onActivitySaveInstanceState(Activity a, Bundle b) { }
            @Override public void onActivityDestroyed(Activity a) { }
        });
    }

    // ── 内部：发送 ─────────────────────────────────────────

    private JSONObject pageContext(Long stay) {
        JSONObject o = new JSONObject();
        try {
            o.put("pgid", curPgid);
            JSONObject pp = TrackEvent.mapToJson(curPgParams == null ? new HashMap<>() : curPgParams);
            if (pp != null && pp.length() > 0) o.put("pg_params", pp);
            if (!refPgid.isEmpty()) o.put("ref_pgid", refPgid);
            if (!refEid.isEmpty()) o.put("ref_eid", refEid);
            if (stay != null) o.put("pg_stay", stay);
        } catch (JSONException ignored) { }
        return o;
    }

    private void emit(String event, JSONObject page, JSONObject element, JSONObject biz) {
        emit(event, page, element, biz, null);
    }

    private void emit(String event, JSONObject page, JSONObject element, JSONObject biz, JSONObject pt) {
        queue.push(TrackEvent.of(event, common.snapshot(), page, element, biz, pt));
    }

    /** API 33 以下无 InputStream.readAllBytes()，华为等旧系统会因此闪退。 */
    private static byte[] readBytes(java.io.InputStream in) throws java.io.IOException {
        java.io.ByteArrayOutputStream buf = new java.io.ByteArrayOutputStream();
        byte[] chunk = new byte[4096];
        int n;
        while ((n = in.read(chunk)) != -1) {
            buf.write(chunk, 0, n);
        }
        return buf.toByteArray();
    }
}

File: sdk/flutter/giso_tracker/lib/src/tracker.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';

import 'package:flutter/foundation.dart';
import 'package:flutter/widgets.dart';
import 'package:http/http.dart' as http;
import 'package:shared_preferences/shared_preferences.dart';

import 'common.dart';
import 'config.dart';
import 'event_queue.dart';
import 'types.dart';

/// GISO tracker facade for Flutter — manual page/element/biz events.
class GisoTracker with WidgetsBindingObserver {
  GisoTracker._();

  static final GisoTracker instance = GisoTracker._();

  GisoConfig? _config;
  SharedPreferences? _prefs;
  EventQueue? _queue;
  String _uid = '';
  String _platform = 'android';
  bool _lifecycleHooked = false;
  int _foregroundTs = 0;

  _PageState? _curPage;
  String _refPgid = '';
  String _refEid = '';

  bool get isInitialized => _config != null;

  /// Initialize once at app start. Call [GisoLifecycleBinding.attach] for lifecycle events.
  Future<void> init(GisoConfig config) async {
    _config = config;
    _prefs = await SharedPreferences.getInstance();
    _platform = config.platform ?? _detectPlatform();
    _queue = EventQueue(
      endpoint: config.endpoint,
      appKey: config.appKey,
      debug: config.debug,
      prefs: _prefs!,
      batchSize: config.batchSize,
      flushIntervalMs: config.flushIntervalMs,
    );
    await _fetchRemoteConfig();
    if (await markActivated(_prefs!)) {
      await _emit('app_install');
    }
    await _emit('app_launch');
  }

  void attachLifecycle() {
    if (_lifecycleHooked) return;
    WidgetsBinding.instance.addObserver(this);
    _lifecycleHooked = true;
  }

  void detachLifecycle() {
    if (!_lifecycleHooked) return;
    WidgetsBinding.instance.removeObserver(this);
    _lifecycleHooked = false;
  }

  @override
  void didChangeAppLifecycleState(AppLifecycleState state) {
    switch (state) {
      case AppLifecycleState.resumed:
        _foregroundTs = DateTime.now().millisecondsSinceEpoch;
        unawaited(_emit('app_foreground'));
      case AppLifecycleState.paused:
      case AppLifecycleState.detached:
        final fgDur = _foregroundTs > 0
            ? DateTime.now().millisecondsSinceEpoch - _foregroundTs
            : 0;
        unawaited(_emit('app_background', pageExtras: {'fg_dur': fgDur}));
        unawaited(flush());
      case AppLifecycleState.inactive:
      case AppLifecycleState.hidden:
        break;
    }
  }

  void setUid(String uid) => _uid = uid;
  void clearUid() => _uid = '';

  Future<void> enterPage(
    String pgid, [
    ParamMap? pgParams,
    Passthrough? pt,
  ]) async {
    if (_curPage != null) await exitPage();
    _curPage = _PageState(
      pgid: pgid,
      params: pgParams,
      pt: pt,
      enterTs: DateTime.now().millisecondsSinceEpoch,
    );
    await _emit('page_enter', pt: pt);
  }

  Future<void> exitPage() async {
    if (_curPage == null) return;
    final stay = DateTime.now().millisecondsSinceEpoch - _curPage!.enterTs;
    final pt = _curPage!.pt;
    await _emit('page_exit', pageExtras: {'pg_stay': stay}, pt: pt);
    _refPgid = _curPage!.pgid;
    _curPage = null;
  }

  Future<void> elementClick({
    required String eid,
    String? mod,
    int? pos,
    ParamMap? params,
    Passthrough? pt,
  }) async {
    _refEid = eid;
    await _emit(
      'element_click',
      element: ElementContext(eid: eid, mod: mod, pos: pos, params: params),
      pt: pt,
    );
  }

  Future<void> elementExposure({
    required String eid,
    String? mod,
    int? pos,
    ParamMap? params,
    required int expDur,
    required double expRatio,
    Passthrough? pt,
  }) async {
    await _emit(
      'element_exposure',
      element: ElementContext(
        eid: eid,
        mod: mod,
        pos: pos,
        params: params,
        expDur: expDur,
        expRatio: expRatio,
      ),
      pt: pt,
    );
  }

  Future<void> bizEvent(
    String code, [
    ParamMap? params,
    Passthrough? pt,
  ]) async {
    await _emit(
      'biz_event',
      biz: BizContext(code: code, params: params),
      pt: _mergePt(_curPage?.pt, pt),
    );
  }

  /// Low-level track for advanced callers.
  Future<void> track(Map<String, dynamic> envelope) async {
    _queue?.push(envelope);
  }

  Future<void> flush() => _queue?.flush() ?? Future.value();

  PageContext _pageContext({Map<String, Object?>? extras}) {
    final pgid = _curPage?.pgid ?? '';
    final base = PageContext(
      pgid: pgid,
      pgParams: _curPage?.params,
      refPgid: _refPgid.isEmpty ? null : _refPgid,
      refEid: _refEid.isEmpty ? null : _refEid,
      pgStay: extras?['pg_stay'] as int?,
    );
    if (extras == null || extras.isEmpty) return base;
    final json = base.toJson();
    for (final e in extras.entries) {
      if (e.key != 'pg_stay') json[e.key] = e.value;
    }
    return PageContext(
      pgid: json['pgid'] as String,
      pgParams: json['pg_params'] as ParamMap?,
      refPgid: json['ref_pgid'] as String?,
      refEid: json['ref_eid'] as String?,
      pgStay: json['pg_stay'] as int?,
    );
  }

  Future<void> _emit(
    String event, {
    ElementContext? element,
    BizContext? biz,
    Map<String, Object?>? pageExtras,
    Passthrough? pt,
  }) async {
    final cfg = _config;
    final queue = _queue;
    final prefs = _prefs;
    if (cfg == null || queue == null || prefs == null) {
      throw StateError('GisoTracker.init() must be called first');
    }
    final common = await buildCommon(
      config: cfg,
      prefs: prefs,
      uid: _uid,
      platform: _platform,
    );
    final envelope = <String, dynamic>{
      'event': event,
      'log_id': newLogId(),
      'ctime': DateTime.now().millisecondsSinceEpoch,
      'common': common.toJson(),
    };
    final page = _pageContext(extras: pageExtras);
    if (page.pgid.isNotEmpty || pageExtras != null) {
      envelope['page'] = page.toJson();
    }
    if (element != null) envelope['element'] = element.toJson();
    if (biz != null) envelope['biz'] = biz.toJson();
    final mergedPt = _mergePt(_curPage?.pt, pt);
    if (mergedPt != null && mergedPt.isNotEmpty) envelope['pt'] = mergedPt;
    queue.push(envelope);
  }

  Future<void> _fetchRemoteConfig() async {
    final cfg = _config;
    final queue = _queue;
    if (cfg == null || queue == null) return;
    final url = cfg.endpoint.replaceFirst(RegExp(r'/v1/track/?$'), '/v1/config');
    if (url == cfg.endpoint) return;
    try {
      final resp = await http.get(Uri.parse(url)).timeout(const Duration(seconds: 5));
      if (resp.statusCode != 200) return;
      final c = jsonDecode(resp.body);
      if (c is! Map) return;
      queue.updateBatching(
        batchSize: (c['batch_size'] as num?)?.toInt() ?? cfg.batchSize,
        flushIntervalMs:
            (c['flush_interval_ms'] as num?)?.toInt() ?? cfg.flushIntervalMs,
      );
    } catch (_) {
      /* silent fallback */
    }
  }

  Passthrough? _mergePt(Passthrough? a, Passthrough? b) {
    if (a == null && b == null) return null;
    final merged = <String, dynamic>{...?a, ...?b};
    return merged.isEmpty ? null : merged;
  }

  String _detectPlatform() {
    if (kIsWeb) return 'web';
    if (Platform.isIOS) return 'ios';
    return 'android';
  }
}

class _PageState {
  _PageState({
    required this.pgid,
    this.params,
    this.pt,
    required this.enterTs,
  });

  final String pgid;
  final ParamMap? params;
  final Passthrough? pt;
  final int enterTs;
}
